File: #0086.partition-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* partition(ListNode* head, int x) {

        ListNode* result = new ListNode(-1);
        result->next = head;
        ListNode* work = result;
        ListNode* cursor = head;
        while (work->next && work->next->val < x) {
             work = work->next;
        }

        cursor = work;
        ListNode* next;
        while (cursor->next) {
            if (cursor->next->val < x){
                next = cursor->next;
                cursor->next = next->next;
                next->next = work->next;
                work->next = next;
                work = next;
            } else {
                 cursor = cursor->next;
            }
        }
        return result->next;
    }
};

File: #0086.partition-list.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} x
 * @return {ListNode}
 */
var partition = function(head, x) {
    if (head == null) {
        return head;
    }
    
		const next = new ListNode();
		const temp = new ListNode();
		next.next = temp.next = head;
		let first = next;
		let curr = temp;
		while (curr.next) {
			if (curr.next.val < x) {
				first = first.next;
				curr.next = curr.next.next;
			} else {
				first.next = first.next.next;
				curr = curr.next;
			}
		}
		first.next = temp.next;
		return next.next;
};
